src/components/tl/ucp/tl_ucp_context.c

```diff
@@ -13,6 +13,106 @@
+    if (test) {                                                                \
+        tl_error(ctx->super.super.lib, msg);                                   \
+        ucc_status = return_status;                                            \
+        goto go;                                                               \
+    }
+
+static inline ucc_status_t
+ucc_tl_ucp_eps_ephash_init(const ucc_base_context_params_t *params,
+                           ucc_tl_ucp_context_t *           ctx,
+                           tl_ucp_ep_hash_t **ep_hash, ucp_ep_h **eps)
+{
+    if (params->context->params.mask & UCC_CONTEXT_PARAM_FIELD_OOB) {
+        /* Global ctx mode, we will have ctx_map so can use array for eps */
+        *eps = ucc_calloc(params->context->params.oob.n_oob_eps,
+                          sizeof(ucp_ep_h), "ucp_eps");
+        if (!(*eps)) {
+            tl_error(ctx->super.super.lib,
+                     "failed to allocate %zd bytes for ucp_eps",
+                     params->context->params.oob.n_oob_eps * sizeof(ucp_ep_h));
+            return UCC_ERR_NO_MEMORY;
+        }
+    } else {
+        *eps     = NULL;
+        *ep_hash = kh_init(tl_ucp_ep_hash);
+    }
+    return UCC_OK;
+}
+
+static inline ucc_status_t ucc_tl_ucp_context_service_init(
+    ucp_params_t ucp_params, ucp_config_t *ucp_config,
+    ucp_worker_params_t worker_params, const ucc_base_context_params_t *params,
+    ucc_tl_ucp_context_t *ctx)
+{
+    ucc_status_t  ucc_status = UCC_OK;
+    ucp_context_h ucp_context_service;
+    ucp_worker_h  ucp_worker_service;
+    ucs_status_t  status;
+
+    if (ctx->cfg.service_worker != 0) {
+        ctx->service.is_used = 1;
+        if (*ctx->cfg.service_tls) {
+            UCP_CHECK(
+                ucp_config_modify(ucp_config, "TLS", ctx->cfg.service_tls),
+                "failed to set UCX_TLS for service worker", err_cfg, ctx);
+        }
+        if (*ctx->cfg.service_devs) {
+            UCP_CHECK(ucp_config_modify(ucp_config, "NET_DEVICES",
+                                        ctx->cfg.service_devs),
+                      "failed to set UCX_NET_DEVICES for service worker",
+                      err_cfg, ctx);
+        }
+        UCP_CHECK(ucp_init(&ucp_params, ucp_config, &ucp_context_service),
+                  "failed to init ucp context for service worker", err_cfg,
+                  ctx);
+
+        UCP_CHECK(ucp_worker_create(ucp_context_service, &worker_params,
+                                    &ucp_worker_service),
+                  "failed to create ucp service worker", err_worker_create,
+                  ctx);
+
+        ctx->service.ucp_context    = ucp_context_service;
+        ctx->service.ucp_worker     = ucp_worker_service;
+        ctx->service.worker_address = NULL;
+
+        CHECK(UCC_OK != ucc_context_progress_register(
+                            params->context,
+                            (ucc_context_progress_fn_t)ucp_worker_progress,
```

Comment:
I also suggest a small change/opt regarding progress. I think we should add 
ucc_tl_ucp_worker_progress function that looks smth like this:
```
{
        ucp_worker_prgoress(ctx->ucp_worker);
        if (ctx->service_worker) {
              int throttling_count = ucc_atomic_fadd32(ctx->service_worker_throttling_count, 1);
              if (throttling_count == ctx->cfg.service_worker_throuttling_count) {
                  ctx->service_worker_throttling_count = 0;
                  ucp_worker_prgoress(ctx->service_worker);
              }
        }
}
```

The idea is that ucp_worker_prgoress is very costly operation (multiple fn calls, cq polling, events polling etc) and we want to minimize its effect. Service team is used only in control path, so most of hte time we can progress it only every so often. We can add smth like "throttling parameter", say = 100. Then service_worker progress is called only once in 100 tl_ucp progresses

Reply:
It is implemented in the commit [TL/UCP: add throttling param for progress fn](https://github.com/openucx/ucc/pull/560/commits/18c4a87658b8be9ea1c38e37aa0be7a16d1f159f)
The parameter is set by the env variable `UCC_TL_UCP_SERVICE_THROTTLING_THRESH`, default=100